Hybrid Planning: Add review changes to feature branch and rebase

sjahr opened the pull request:
### Description

This PR reopens #718 with the intention to get the hybrid planning feature merged into main. 
I've rebased again on main (11/18/2021) and added the changes introduced by #763 (Thanks to @tylerjw and @AndyZe!). Furthermore, I've fixed some things to make the demo work again, although that is still work in progress right now. 
@AndyZe and @tylerjw I've sent you collaborator invitations to my moveit2 fork, so you can directly contribute to this PR to speed up the process.

### Checklist
- [X] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Adress review from #488 completely
- [ ] Fix segfault caused by invalid action goal handle in hybrid planning interface

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

sjahr commented on moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/src/hybrid_planning_interface.cpp:
```diff
@@ -0,0 +1,242 @@
+
+  // Forward global trajectory goal from hybrid planning request TODO(sjahr) pass goal as function argument
+  auto global_goal_msg = moveit_msgs::action::GlobalPlanner::Goal();
+  global_goal_msg.motion_sequence =
```
This causes a segfault right now

AndyZe commented:
I appreciate you opening a new PR from a branch I have access to. I'm pretty deep into addressing the code reviews on a branch of my own. I plan to push onto this branch soon. Sound good?

sjahr commented:
> I appreciate you opening a new PR from a branch I have access to. I'm pretty deep into addressing the code reviews on a branch of my own. I plan to push onto this branch soon. Sound good?

Sounds perfect thanks!

sjahr commented:
@AndyZe FYI, I've marked the review comments in the other PR that I addressed with this PR with :+1: to keep an overview

sjahr commented:
@AndyZe FYI I've marked the review comments that I've addressed so far with :+1:  in the old PR

AndyZe commented:
@sjahr I'm sorry but rebasing my branch on this one was painful and the demo was crashing. I don't know why. I went ahead and opened #813 instead. Hopefully we can get that one merged soon. Will close this in a bit unless you have an objection.

sjahr commented:
> @sjahr I'm sorry but rebasing my branch on this one was painful and the demo was crashing. I don't know why. I went ahead and opened #813 instead. Hopefully we can get that one merged soon. Will close this in a bit unless you have an objection.

@AndyZe no worries, my only concern is that you are doing some duplicate work as I've addressed a lot of the review comments in #718/  #813 but I understand that it might take you more time to unify both branches instead of simply fixing those by yourself.

I've mostly addressed #814 with [a985b8c](https://github.com/ros-planning/moveit2/pull/811/commits/a985b8c3fc74f75551e74159e4849936bba55067), [b027f62](https://github.com/ros-planning/moveit2/pull/811/commits/b027f625442b7f53091ff3b7af6f26bae7a3e9fd) and [5682263](https://github.com/ros-planning/moveit2/pull/811/commits/5682263a4fec2fe2f5efa404cec2f2f263e1c111) of this PR. I've removed the timers and added another member class to the hybrid planning manager which contains the ros2 communication interfaces (publisher, subscriber etc). Thereby, I can give the planning logic access to those, without the need of passing a shared pointer of the whole hybrid planning manager to the logic. The only thing that does not work right now, is the thing I've pointed out [in a comment in this PR](https://github.com/ros-planning/moveit2/pull/811#discussion_r752319895): After successfully moving it in the member variable in line 214, the goal_handle becomes invalid when next used in line 141. 

I think the first commit of the 3 mentioned above can be cherry-picked without concern, because the segfault I've experienced is caused in the hybrid_planning_manager after adding the latter two.

henningkayser commented:
> I appreciate you opening a new PR from a branch I have access to. I'm pretty deep into addressing the code reviews on a branch of my own. I plan to push onto this branch soon. Sound good?

Can we get this huge chunk of code merged into the feature branch first? It's much easier to follow what's happening if there are not plenty of different authors pushing to multiple PR branches

AndyZe commented:
Let's get this one merged instead:  https://github.com/ros-planning/moveit2/pull/813

@sjahr thanks, I will check about cherry-picking a985b8c
